pandas/util/_decorators.py

```diff
@@ -275,7 +275,8 @@ def wrapper(*args, **kwargs) -> Callable:
             elif hasattr(arg, "_docstr_template"):
                 templates.append(arg._docstr_template)  # type: ignore
             elif arg.__doc__:
-                templates.append(arg.__doc__)
+                doc_tmp = arg.__doc__.replace("{", "{{").replace("}", "}}")
```

Comment:
I don't understand this, do you mind explaining why is this needed?

Reply:
Hi, a short quick answer: 
This change solves the error when we want to reference the docstring from function which not using doc decorator.

You must already know this, but just want to clarify this again so that helps others catch up:
When we format string in Python, we use `{` and `}` as a place holder, and use the arguments to replace the place holder. For example, `'Hello {name}!'.format(name='Tom')` will become `'Hello Tom!'`. However, if we want to show `{` or `}` from a formatted string, we have to put `{{` and `}}`. For example, `'Hello {{name}}!'.format(name='Tom')` will become `'Hello {Tom}!'`. 

Back to our case: 
In the docstring, `{` and `}` are very commonly used to represent set. The old implementation will just take the original docstring, and then `{` and `}` will be considered as a place holder. This will cause an error. We can fix that by switching `{` and `}` to `{{` and `}}` in the doc decorator. Then, it will be shown as `{` and `}`.

Please take the changes in [this file](https://github.com/pandas-dev/pandas/blob/a1b0e0bfd1a1c8bf65b69f5acf298390da57dd02/pandas/core/indexing.py) as an example:
The docstring of `IndexingMixin.iloc` has `{` and `}`, these changes will allow us to do something like [this](https://github.com/pandas-dev/pandas/blob/a1b0e0bfd1a1c8bf65b69f5acf298390da57dd02/pandas/core/indexing.py#L1346):
```python
@doc(IndexingMixin.iloc)
class _iLocIndexer(_LocationIndexer):
```

I am not sure if I explain the reason clear enough, please feel free to come with follow up questions.

Reply:
Sorry, I still don't get it. I see these cases:

**Not using `@doc`**:
- Enumeration: `{'a', 'b'}`
- Variable: Doesn't apply

**Using `@doc`**:
- Enumeration: `{{'a', 'b'}}`
- Variable: `{name}`

I guess what this solves is taking a docstring from a non-decorated function, and use it with the decorator, converting it from the first case to the second. But this feels a bit hacky, and I'm not sure if it would make more sense to apply `.format()` only when it makes sense instead. I'm a bit worried on making this too complex and tricky, and having unexpected side-cases.

And if we need to keep this implementation, we need to add comments so a reader can understand what's going on, and what these `if` cases mean.

Reply:
> I guess what this solves is taking a docstring from a non-decorated function, and use it with the decorator, converting it from the first case to the second.

Yes, exactly! Thanks for helping me explain this. That is want I try to say.

> ... I'm not sure if it would make more sense to apply .format() only when it makes sense instead.

That make sense to me. I would like to give it a try. 

> I'm a bit worried on making this too complex and tricky, and having unexpected side-cases.

I agree with you here. I would like to try the idea above first, and a backup solution might be using string template to avoid this case.

Reply:
>  ... I'm not sure if it would make more sense to apply .format() only when it makes sense instead.

Hi @datapythonista, I have modified the `@doc` to apply `.format()` only when it is coming from decorated function or docstring.

It seems there is some issue with the CI checks. I will keep working on this, but also very welcome to get some help here.